### ISSUE TITLE
Type annotation for all exports in the module

### ISSUE THREAD

**Issue by mohsen1** [microsoft/TypeScript#38511]
## Search Terms
Export, typed exports, modules, Next.js, Redwood.js 

## Suggestion
Allow a new syntax that will allow typing all of exports in a module.

```ts
type export = {
  default: string
  foo: (bar: boolean) => number
}

export default "Hello world!";
export function foo(bar) { return 1 }
```

### Sharing "module types"
It should be possible to import module types

```ts
import ModuleType from './module.ts';

type export = ModuleType;
```

### Alternative syntaxes 

```ts
module type = {}
```

```ts
export module {}
```

## Use Cases and Examples
Increasingly frameworks are using named module exports as a way of organizing their code. For instance [Next.js](https://nextjs.org/docs/basic-features/data-fetching) is using named exports to specify things like exported component and functions for fetching data:

```ts
function Blog({ posts }) {
  return (
    <ul>
      {posts.map(post => (
        <li>{post.title}</li>
      ))}
    </ul>
  )
}


export async function getStaticProps() {
  return {posts: []}
}

export default Blog
```

### Frameworks relying on named export 

* Next.js
* Redwood.js
* > please suggest more 

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### Related issues

* https://github.com/microsoft/TypeScript/issues/13626

**Comment by balazsorban44** [comment]
You can add Remix to this category

https://blog.remix.run/p/remix-preview

It will export a load function similar to Next.js's getServerSideProps

**Comment by kentcdodds** [comment]
I didn't notice this issue when I opened: https://github.com/microsoft/TypeScript/issues/46942

There are a few more details for use cases and an alternative syntax, but I don't really care too much about the specific syntax. I just want the feature. :) You've got a 👍 from me!

**Comment by Rich-Harris** [comment]
Adding [SvelteKit](https://kit.svelte.dev) to the list of frameworks that follows this pattern — this would be hugely beneficial to us as well.

An even better outcome for users would be if modules could be typed implicitly — for example if certain files (either with a particular naming convention or because the framework generates a config that lists them) were known to implement a particular interface, the `export implements` (or whatever) would be unnecessary (unless you needed to specify type variables). For example a typical SvelteKit [endpoint](https://kit.svelte.dev/docs#routing-endpoints) looks like this:

```ts
// src/routes/echo.ts
import { RequestHandler } from '@sveltejs/kit';

export const get: RequestHandler = (request) => {
  return {
    status: 200,
    headers: {},
    body: JSON.stringify(request)
  };
}
```

Converting that to an explicitly typed module...

```diff
// src/routes/echo.ts
-import { RequestHandler } from '@sveltejs/kit';
+import { Endpoint } from '@sveltejs/kit';
+
+export implements Endpoint;

-export const get: RequestHandler = (request) => {
+export function get(request) {
  return {
    status: 200,
    headers: {},
    body: JSON.stringify(request)
  };
}
```

...would only really start paying dividends once you had multiple exports, whereas an implicitly-typed module would provide significant benefits even to people who see the red squigglies but otherwise have no idea what TypeScript is:

```diff
// src/routes/echo.ts
-import { RequestHandler } from '@sveltejs/kit';

-export const get: RequestHandler = (request) => {
+export function get(request) {
  return {
    status: 200,
    headers: {},
    body: JSON.stringify(request)
  };
}
```

**Comment by kentcdodds** [comment]
Hey @Rich-Harris 👋

I'd also love implicit types based on a convention and the tsconfig-based configuration of that convention (😆). That would be a big benefit for Remix users as well.

I'm concerned that this would stall the feature. If there's some way to have both and start with the explicit that's the future I would prefer personally. I think explicit has a place in the TS ecosystem as well, so both make sense to me. So if we could get the explicit finished and start benefitting from that then work on the implicit that would be super :)

**Comment by Rich-Harris** [comment]
Indeed. If feature B depends on feature A, it's better to implement feature A first rather than waiting to implement both.

Nevertheless, it's very useful in my experience to understand the rough shape of feature B (and any other dependent features you have in mind) _before_ designing feature A, lest you end up with a design that for whatever reason makes those follow-ups hard or impossible. For that reason it's often worthwhile to include those considerations in the initial discussion.

**Comment by GregBrimble** [comment]
+1 from Cloudflare Pages as well!

We include declare this `PagesFunction` type in our [@cloudflare/workers-types]() package, and encourage people to include it in their `types` in `tsconfig.json`:

```json
{
  "compilerOptions": {
    "types": ["@cloudflare/workers-types"]
  }
}
```

With support for type-checking of modules, we could declare a `PagesFunctions` or similar:

```ts
export implements PagesFunctions

export const onRequestPost = async ({ request }) => {
  const { name } = await request.json()
  return new Response(`Hello, ${name}!`)
}

export const onRequestGet = () => {
  return new Response("Hello, user!")
}
```

But this API would do exactly what we'd need! Thanks everyone for suggesting it :)

**Comment by geelen** [comment]
@Rich-Harris love your idea of implicit naming based on file pattern, imo in tsconfig something like:

```js
{
  "compilerOptions": {
    "exportTypings: {
      "*.svelte.ts": "@sveltejs/kit::Endpoint"
    }
  }
}
```

Maybe with something less c++-y than `::` but you get the idea.

I think `<module-name>.<exports-types>.ts` could become a fairly common convention. But equally, for all-one-type projects you could just set `*.ts` to be any type you like, or for Pages we could set a subdir like: 

```js
"functions/**/*.ts":"@cloudflare/workers-types::PagesFunctions"
```

I'm liking this idea more and more...

**Comment by orta** [comment]
Notes from the last time we talked about this: https://github.com/microsoft/TypeScript/issues/38713 and some of the possible consensus https://github.com/microsoft/TypeScript/issues/420#issuecomment-632338084

**Comment by kentcdodds** [comment]
Thanks @orta. Looks like there's community interest and team appetite for the idea. So where do we go from here? How do we push things forward?

**Comment by shilman** [comment]
Mentioned in other threads on the topic, Storybook's [Component Story Format](https://storybook.js.org/blog/component-story-format/) also heavily relies on this pattern, where the default export includes metadata about a component, and each named export is a component example, or **story**:


```ts
import { Meta, Story } from '@storybook/react';
import { Button } from './Button';

const meta: Meta = {
  title: 'Demo/Button',
  component: Button;
};
export default meta;

export const Primary: Story = () => <Button primary>Primary</Button>;

export const Disabled: Story = () => <Button disabled>Disabled</Button>;
```

In our case, users also want more type safety and better autocompletion than the general types provide, so they will parameterize them by the props of the component that is being documented. So at the top of each file they might make specific versions at the top of each file:

```ts
type Meta = ComponentMeta<typeof Button>;
type Story = ComponentStory<typeof Button>;

const meta: Meta = { ... };
...
```

Ideally a solution would be powerful enough to express this kind of specificity.

So this would fit the bill in each file:

```ts
export implements CSF<typeof Button>;
```

But this proposal in its current form would not:

```ts
"src/**/*.stories.tsx":"@storybook/react::CSF"
```

A combination of the former (for users who want more control) and the latter (for convenience) would be amazing. 💯

**Comment by hollandThomas** [comment]
> Nevertheless, it's very useful in my experience to understand the rough shape of feature B (and any other dependent features you have in mind) _before_ designing feature A, lest you end up with a design that for whatever reason makes those follow-ups hard or impossible. For that reason it's often worthwhile to include those considerations in the initial discussion.

In this spirit, I just wondered about the following as a possible future extension of this proposal:
_A module that implements named exports may infer the types of any **imports** that originate from the same module that provides the named export interface._

I am not sure about the semantics and _how_ to implement this but maybe someone can come up with something smart.

Taking Remix as an example, this would enable e.g.

```typescript
// Internal Remix code
interface EntryRouteModule<LoaderData = Response | AppData> {
  CatchBoundary?: CatchBoundaryComponent;
  ErrorBoundary?: ErrorBoundaryComponent;
  default?: RouteComponent;
  handle?: RouteHandle;
  links?: LinksFunction;
  meta?: MetaFunction | HtmlMetaDescriptor;
  action?: ActionFunction;
  headers?: HeadersFunction | { [name: string]: string };
  loader?: LoaderFunction<LoaderData>;
}
```

Then, in a route

```typescript
// routes/posts.tsx
import type { EntryRouteModule } from 'Remix'
import { useLoaderData } from 'Remix';

export implements EntryRouteModule<{ posts: Post[] }>;

// this is typed correctly because of the line above
export const loader = async () => {
  return {
    posts: await postsQuery()
  }

export default function Blog() {
  // Have TypeScript figure this out automatically somehow:
  // useLoaderData now knows that `LoaderData === Product[]`. There is no need to `useLoaderData<Post[]>`
  const { posts } = useLoaderData()

  return (
    <ul>
      {posts.map(post => (
        <li>{post.title}</li>
      ))}
    </ul>
  )
}
```

This is a slightly different use case than the one outlined in the initial Next.js example, where the inference of `posts === Post[]` does not rely on any additional imports.

```typescript
import { PageModule } from 'next'

type Post = { title: string };
export implements PageModule<{ posts: Post[] }>;

// Here, `posts` is being inferred correctly as per proposal
function Blog({ posts }) {
  return (
    <ul>
      {posts.map(post => (
        <li>{post.title}</li>
      ))}
    </ul>
  )
}

export async function getStaticProps() {
  return {posts: []}
}

export default Blog
```

**Comment by mohsen1** [comment]
@DanielRosenwasser just let me know that [it's possible to import self](https://github.com/microsoft/TypeScript/issues/47151#issuecomment-1015975676) to simulate what we're asking for. Just in case you're curious this is how it would look like in Next.js for instance:

```ts
// @file pages/posts.tsx

import React from 'react';

type InferGetStaticPropsType<T> = T extends () => Promise<{ props: infer U }> ? U : never;

const Blog: React.FC<InferGetStaticPropsType<typeof import('./posts').getStaticProps>> = ({
  posts,
}) => {
  return (
    <ul>
      {posts.map((post) => (
        <li>{post.title}</li>
      ))}
    </ul>
  );
};

export async function getStaticProps() {
  const posts: Array<{ title: string }> = [];
  return { props: { posts } };
}

export default Blog;
```

This is less automatic than typing all exports using something like `export implements` but it's possible to do today.

**Comment by ivanhofer** [comment]
I would also love to see such a feature in TypeScript.
Probably every project that has file conventions would benefit from such a feature. My project [`typesafe-i18n`](https://github.com/ivanhofer/typesafe-i18n) could also benefit from it.

I don't think that there is a new syntax needed. What if TypeScript could use a `.d.ts` file to check what variables and exports a file should have?

Currently e.g. a `file.d.ts` would tell every other file that imports `file.js` what values it exports and what shape those variables and functions have. But in that `file.js` no type-check happens. The `file.js` could also export something totally different and TypeScript would **not** throw an error:

_file.d.ts_
```ts
declare export someVariable: number
```
_file.js_
```js
export someVariable = 'Hello'
```

So if TypeScript would use the information from the `d.ts` file to check if the `.js` file was implemented correctly, we could have a solution to the typed export problem. Of course this should also work with `.ts` files.

With a solution like we can see in SvelteKit where types are generated and then linked with the real source files via the `rootDirs` compiler option, those `.d.ts` files could be generated without adding declaration files inside the actual `src` folder.

> Having those checks enabled would probably break a few existing code bases if they manually create `.d.ts` files. Or actually they would get notified that they defined something wrong ;)
> But this is a topic for a later discussion. I just want to propose an alternative to this thread.

**Comment by Rich-Harris** [comment]
That's a stroke of genius @ivanhofer — it solves the problem very elegantly. I've often wondered why `.d.ts` files don't already work that way.

My conclusion was that TypeScript probably assumes that `foo.d.ts` is generated from `foo.ts` and as such represents a snapshot of its declarations the last time `tsc` was run, but even though emitting `.d.ts` files as siblings to their inputs is the default behaviour if `"declaration"` is true and `"declarationDir"` is unset, I've never seen a case where it's actually _desirable_.

I wonder if a new compiler option like `"selfDeclaration": true` could solve this problem?

**Comment by ivanhofer** [comment]
Yes usually the `.d.ts` should be genereated from the actual source file. But there exist a lot of libraries that are written in plain JavaScript and types are added by manually creating the `.d.ts` file or there are also all the [`@types`](https://www.npmjs.com/~types) packages were the types come from a 3rd party.

Using the declaration files to type the corrresponding source file makes only sense for your own source files (everything that matches "include" from your `tsconfig.json`).
_The more I think about this topic, the less I understand the [`skipLibCheck`](https://www.typescriptlang.org/tsconfig#skipLibCheck) option. Why should I want TypeScript to check files from my `node_modules` folder I don't use in my code? (maybe I also don't fully unserstand what it really does)_

A compiler option where someone could opt-in into the `.d.ts` typechecking behaviour would be a good solution. I could also immagine this beeing one of the "strict" options since it potentially warns you about big mistakes in your code base.

Such a feature would also make it possible to have "scoped global types" (not sure how to call them). A concrete example for SvelteKit would be the `GET` type you can use to type endpoints. Currently you have to manually import the type and also be sure you import the generated type and not the generic type. With the `.d.ts` file the `GET` type could be "just there". No need to import it. Like the interface `HTMLElement` can be used everywhere without importing it directly. But in this case the type would be scoped to a single file.

**Comment by camjackson** [comment]
@ivanhofer With your suggested solution, would that I mean I need to create a `.d.ts` file for every module that I want to type? So with Remix as an example, for every single one of my route `.ts` files I would need to create a corresponding `.d.ts` file, all of which would have identical contents (seeing as all route modules follow the same pattern)?

**Comment by ivanhofer** [comment]
@camjackson _I haven't used Remix yet, so I don't fully know how route files work there. I just looked at some examples in the comments above._
If you want all your route files beeing typed without a user having to import the correct type, then with the solution I have suggested probably yes.
But those files don't have to contain a dozen of lines. I can imagine having a file where you define how a route should look like and then just writing `export * from '@framework/common-route'` for all other files.
Having those files located in your normal `src` folder is probably not a good user experience. I really like the solution of auto-generating and hiding those files via the [`rootDirs`](https://www.typescriptlang.org/tsconfig#rootDirs) option like SvelteKit does.

**Comment by camjackson** [comment]
Yeah so the motivating example for many of the frameworks mentioned (Remix, Next, Svelte) is that app developers create a whole bunch of files which all share the same export convention. I can't speak for everyone but personally I'm totally fine with adding a line or 3 to each relevant file, which states "this is a remix route module".

What I _don't_ want, is to have to pepper my src folder with identical `.d.ts` files. A little bit of boilerplate in each file is fine, but a bunch of small, boilerplate-only files would be too much noise IMO.

I like the suggestion from the opening comment on this issue:

```
import { PageModule } from 'next'

type Post = { title: string };
export implements PageModule<{ posts: Posts[] }>;
```

That's a pretty good illustration of the use case.

**Comment by ivanhofer** [comment]
You can do almost the same right now with a helper function. No need to have a new Syntax.

_adaption of the initial example:_

_some exposed file from the framework_
```tsx
// TODO: specify the correct types
type CreatePageModuleProps<T> = {
	component: (...args: any[]) => JSX.Element,
	getStaticProps: (...args: any[]) => any
	// ... other "exports" the page module should have
}

// since probably only the framework uses the pageModule this object could also look different
// this example just tries to replicate the current structure as close as possible
const createPageModule = <T,>(props: CreatePageModuleProps<T>) => {
	const {component, ...rest} = props
	return Object.assign(component, rest)
}
```

_route file_
```tsx
import { createPageModule } from '@framework/pageModule'

type Post = { title: string };

function Blog({ posts }) {
  return (
    <ul>
      {posts.map(post => (
        <li>{post.title}</li>
      ))}
    </ul>
  )
}

async function getStaticProps() {
  return { props: { posts: [] } }
}

// TypeScript tells you here if you have specified something wrong
export default createPageModule<{ posts: Post[] }>({
	component: Blog,
	getStaticProps
})
```

_usage_
```tsx
import PageModule from './pageMudule'

// the component itself is the default export
const renderedPage = PageModule()

// note: no direct import of { getStaticProps } but this is anyway hidden from the user because only the framework uses this code
const { getStaticProps } = PageModule
```

So instead of exporting each function individually a framework could choose to require to call a provided function that allows to check for type errors.

Like with the `export implements` syntax, if a user forgets to use this line of code, no typechecks will happen.

By having the possibility to use `.d.ts` a user cannot forget to write the line of code that adds typechecks. It is already here, abstracted into in another file. If I would have to manually create those files and add boilerplate code, I also would't like this. In my mind the beauty of this solution is: you can generate and hide the `.d.ts` file from the actual source code the user is writing.

**Comment by geelen** [comment]
> With your suggested solution, would that I mean I need to create a .d.ts file for every module that I want to type?

Surely there's a halfway point here? I think @ivanhofer's thought of using `.d.ts` as an assertion format is absolutely genius, but I agree having to have one-per-input file is cumbersome. So why not apply it a different way?

```json5
// tsconfig.json
{
  compilerOptions: {
    declarationValidations: {
      "**/*.route.ts": "@remix/type-assertions/route.d.ts",
      "**/*.model.ts": "./models/_type.d.ts"
    }
  }
}
```

What's nice is that you could _also_ do what @ivanhofer's suggesting if you allow a bit more syntax in the tsconfig:

```json5
declarationValidations: {
  "src/**/*.ts": "src/$1/$2.d.ts",
  // or, if file-by-file declarations were being generated in a hidden directory:
  "src/**/*.ts": "./svelte/decls/$1/$2.d.ts"
}
```

This, to me, feels shippable & ergonomic. Any use-cases where it breaks down?

**Comment by camjackson** [comment]
> So instead of exporting each function individually a framework could choose to require to call a provided function that allows to check for type errors.

I suggested this on the Remix discord and it was explained to me that it won't work. In Remix, route modules include both frontend and backend code, and the framework needs to be able to tree shake / code split the code apart. So its conventions must be export-based, they can't be fields of a single object.

**Comment by ivanhofer** [comment]
> Surely there's a halfway point here? I think @ivanhofer's thought of using `.d.ts` as an assertion format is absolutely genius, but I agree having to have one-per-input file is cumbersome. So why not apply it a different way?

I agree, having a config option to choose the behaviour where `.d.ts` files get applied would be an extension to the solution that makes sense.
The reason why I proposed this soultion with exactly that state is: for me it looks like it is already implemented in the code base. Just not enabled. I don't want to offend any TypeScript maintainer but for me (without ever looking at the code) it just feels like removing an if statement somewhere or adding two lines of code to enable that feature.

> I suggested this on the Remix discord and it was explained to me that it won't work.

Well, from a technical perspective Remix could also transform the file and split it up into distinct files during the build process. Maybe a bit overkill, but it could already work with the current TypeScript version.
_"Where there is a will, there’s a way" :)_

**Comment by narutosstudent** [comment]
This would be HUGE :fire:

**Comment by nickmccurdy** [comment]
This could also be helpful for globals that are only exposed in certain files, like Node environment variables and test frameworks.

**Comment by mohsen1** [comment]
3 questions:

1. should we allow multiple `export implements` and treat it like an interface extending multiple other interfaces?
2. if `exports` is acting like a type identifier then should `typeof exports` be a valid syntax?
3. Are you accepting pull requests for this?

**Comment by kentcdodds** [comment]
As an interested user, here are my opinions:
1. Maybe? Personally I'd just do `export implements Type1 & Type2`
2. Hmmm... I don't know what use cases that would be useful for 🤔
3. I hope they are!

**Comment by slorber** [comment]
Next.js 13 solution to mitigate for the lack of this feature (+ more):
- TypeScript language service plugin: https://github.com/vercel/next.js/blob/canary/packages/next/server/next-typescript.ts
- Demo: https://twitter.com/delba_oliveira/status/1592209522281439232

<img width="622" alt="CleanShot 2022-11-15 at 10 56 57@2x" src="https://user-images.githubusercontent.com/749374/201889424-54bb2cc6-9910-4e12-9422-86fecee54ee9.png">

**Comment by mohsen1** [comment]
@kentcdodds `typeof exports` can be useful to export the type of all exports in a file for some use:

```ts
export const foo = 1
export const obj: { bar?: string } = {}

export type Exports = typeof export; // { foo: number; obj: { bar?: string } }
```

I don't have an immediate use case for it but I'm sure this can be useful

**Comment by kentcdodds** [comment]
Handy trick to help error if you're exported types are wrong. However, it doesn't help with auto-complete not will it report errors in a convenient location.

**Comment by ggoodman** [comment]
I'd like to add another use-case for this feature. We use `monaco-editor` and its typescript language services to provide an in-browser, IDE-like editing experience for [Auth0 Actions](https://auth0.com/docs/customize/actions/flows-and-triggers/login-flow#access-control). Actions are CJS modules that we execute in a 'serverless' environment on behalf of our customers at specific lifecycle events in our auth pipelines. It is conceivable that we could move to EJS modules if that would be the gateway to having access to this feature if it came to TypeScript.

Currently, we rely on seeding Actions with skeleton templates that use jsdoc annotations to power the intellisense. This _works_ but means that the majority of the template is consumed by non-functional boilerplate. Supporting typed modules such as in this proposal would give us (and I suspect all other serverless-like platforms / tools) an improved way to deliver a more robust and elegant DX for our customers / users.

**Comment by ggoodman** [comment]
One thing I don't like so much about the approaches that involve new syntax is that it requires author of the subject module to _opt in_. From what I understand of the use-cases that motivate this issue, the types for module exports are not _optional_. Instead, these types are being _enforced_ by an external factor or framework. I think it would be interesting if the enforcement could thus come from outside too. 

From that angle, I wonder if it might be interesting to think about a design where these module-level type signatures are 'attached' to files through the `tsconfig.json` and equivalent language service constructs.

A straw-man idea follows. Here, we can define a set of mappings from files globs to a module specifier and exported symbol that defines the expected shape of the matching files' exports.

```json
{
  "exportTypes": [{
    "include": ["**/pages/*.ts"],
    "typePath": "@framework/types",
    "typeName": "PageExports",
  }],
}
```

In other words, any file matching the glob `**/pages/*.ts` would have the signatures of its exported symbols checked against the `PageExports` type exported from the `@framework/types` module. The type checker and language server would be able to enforce the types at compile-time and suggest them at authoring time, respectively.

The side-benefit is that affected files no longer need to include superfluous syntax to benefit from this feature. Many frameworks area _already_ providing default `tsconfig.json` settings from which applications `extends` so I could imagine this slotting nicely into that sort of setup.

**Comment by kentcdodds** [comment]
I like that. Though in Remix at least, you can very dynamically define what files count as routes so I would like an additional capability to define a type for the exports.

I'm a fan of what @jamiebuilds-signal suggests with the satisfies keyword.

**Comment by camjackson** [comment]
> From that angle, I wonder if it might be interesting to think about a design where these module-level type signatures are 'attached' to files through the tsconfig.json and equivalent language service constructs.

I see the benefit of that, but I think it makes the code much less *discoverable*. To me it's very important that a new developer can look at a piece of code, without any context, and be able to figure out what it does by following the breadcrumbs.

I prefer an API that explicitly establishes a link that says "this code that you're looking at implements this interface", and then you can click through to see what that interface is. As opposed to it being configured somewhere else, with no way of like, naturally discovering it from the code in question. It becomes something that you just have to know.

**Comment by ggoodman** [comment]
> I prefer an API that explicitly establishes a link that says "this code that you're looking at implements this interface", and then you can click through to see what that interface is.

That's a great point. Perhaps the relevant comparison here is what you have defined in your `tsconfig.json` for the `compilerOptions.lib` option. This is something defined on the 'outside' that affects which ambient types will show up. For use-cases where you want to be explicit, you can opt into a [triple-slash directive](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-lib-) to make the dependency explicit.

**Comment by btoo** [comment]
allowing ambient module declarations to specify relative module names as an alternative to using tsconfig could be a more intuitive and/or powerful way to achieve this by way of using typescript rather than json
```typescript
declare module './MyPage' {
  export {} satisfies RemixRoute<{ bookId: string }>
}
```
and this could work with wildcards as well
```typescript
declare module '**/pages/*.ts' {
  export {} satisfies PageExports
}
```

**Comment by ggoodman** [comment]
Oh nice, @btoo that's very clever!

**Comment by mohsen1** [comment]
I like `export {} satisfies Type` because `satisfies` works better than `implements` in this case. The only thing is the emitted `export {}` in runtime space that might not be desired.

**Comment by EloB** [comment]
@btoo Is that doable at the moment or is that a a feature request?

**Comment by mariusGundersen** [comment]
I would really like this too, and would be happy with `export {} satisfies Type`, although some magic with tsconfig and wildcard paths would be very useful.

A workaround while we wait for this feature is to import the module into itself and then check that what is imported satisfies some type. For example:

```ts
// myModule.ts
export default function () {
  return "test";
}

import * as __SELF__ from './myModule';
__SELF__ satisfies { default(): string };
```
This is quite cumbersome though, since you need to specify the name of the file in the file, and you get the error in the satisfies line, not on the export line.

**Comment by nfroidure** [comment]
Would be nice to be able to create some `.d.ts` file with this:
```ts
declare module 'src/commands/*.ts' satifies/extends {
  name: string;
  definition: CommandDefinition;
  default: Command;
}
```

A bit like https://github.com/vercel/next.js/blob/canary/packages/next/image-types/global.d.ts but for module validation.

**Comment by juancampa** [comment]
We've been following this issue for years (from [membrane.io](https://www.membrane.io)). It would greatly simplify how our users interact with the platform.

Feel free to add us to the list 🙏

**Comment by Brennername** [comment]
> [@Rich-Harris](https://github.com/Rich-Harris) love your idea of implicit naming based on file pattern, imo in tsconfig something like:
> 
> {
>   "compilerOptions": {
>     "exportTypings: {
>       "*.svelte.ts": "@sveltejs/kit::Endpoint"
>     }
>   }
> }
> Maybe with something less c++-y than `::` but you get the idea.

I do not mind using the :: "operator" because that is common in java method references as well, and it clearly indicates you are dealing not with a call using the dot operator but a reference using the :: operator.

But as I read this I was thinking "Why is this _not_ a feature?"

**Comment by bgub** [comment]
> ```js
> // tsconfig.json
> {
>   compilerOptions: {
>     declarationValidations: {
>       "**/*.route.ts": "@remix/type-assertions/route.d.ts",
>       "**/*.model.ts": "./models/_type.d.ts"
>     }
>   }
> }
> ```

Just following up on this issue! My favorite solution is @geelen's proposal (quoted example above). This would be incredibly useful and I'd be happy to help contribute.

**Comment by AgentEnder** [comment]
Throwing Nx's name in the mix as frameworks / tools that would benefit from this
